### PR TITLE
 Refactoring models.py & Adding to Documentation

### PR DESCRIPTION
**Pull request by pruksmhc**
Refactoring long functions in models.py, and also adding documentation on models and FAQs to make getting started and understanding the codebase easier for newcomers.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -132,9 +138,15 @@ tensorboard --logdir <exp_dir>/<run_name>/tensorboard
 
 `--remote_log` (or `-r`): use this to enable remote logging via Google Stackdriver. You can set up credentials and set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable; see [Stackdriver Logging Client Libraries](https://cloud.google.com/logging/docs/reference/libraries#client-libraries-usage-python).
 
-## Model
+## Models (and how to add a Model)
+
+The core model is a shared BiLSTM with task-specific components. When a language modeling objective is included in the set of training tasks, we use a bidirectional language model for all tasks, which is constructed to avoid cheating on the language modeling tasks. We also provide bag of words and RNN sentence encoder. 
 
-The core model is a shared BiLSTM with task-specific components. When a language modeling objective is included in the set of training tasks, we use a bidirectional language model for all tasks, which is constructed to avoid cheating on the language modeling tasks.
+<<<<<<< HEAD
```
Comment (W4ngatang) [inline README.md]: I think you forgot to clean this up?

File: README.md [hunk 2]
```diff
@@ -170,18 +182,18 @@ To add new tasks, you should:
 
 2. Create a class in ``src/tasks.py``, and make sure that...
 
-    * You decorate the task: in the line immediately before ``class MyNewTask():``, add the line ``@register_task(task_name, rel_path='path/to/data')`` where ``task_name`` is the designation for the task used in ``pretrain_tasks, target_tasks`` and ``rel_path`` is the path to the data in ``data_dir``. See `EdgeProbingTasks` in [`tasks.py`](src/tasks.py) for an example.
+    * You decorate the task: in the line immediately before ``class MyNewTask():``, add the line ``@register_task(task_name, rel_path='path/to/data')`` where ``task_name`` is the designation for the task used in ``pretrain_tasks, target_tasks`` and ``rel_path`` is the path to the data in ``data_dir``. See `EdgeProbingTasks` in [`tasks.py`](src/tasks/tasks.py) for an example.
```
Comment (W4ngatang) [inline README.md]: I think the path should be `src/tasks` (edge-probe tasks are in a separate file)

File: README.md [hunk 3]
```diff
@@ -170,18 +182,18 @@ To add new tasks, you should:
...
     * Your task implements a method ``task.get_sentences()`` that iterates over all text to index in order to build the vocabulary. For some types of tasks, e.g. ``SingleClassificationTask``, you only need set ``task.sentences`` to be a list of sentences (``List[List[str]]``).
-    * Your task  implements a method ``task.count_examples()`` that sets ``task.example_counts`` (``Dict[str:int]``): the number of examples per split (train, val, test). See [here](https://github.com/jsalt18-sentence-repl/jiant/blob/master/src/tasks.py#L647) for an example.
+    * Your task  implements a method ``task.count_examples()`` that sets ``task.example_counts`` (``Dict[str:int]``): the number of examples per split (train, val, test). See [here](https://github.com/jsalt18-sentence-repl/jiant/blob/master/src/tasks/tasks.py#L647) for an example.
```
Comment (W4ngatang) [inline README.md]: nit: can you fix the double spaces between "Your task" and "implements" ?

File: src/models.py [hunk 4]
```diff
@@ -538,6 +554,8 @@ def build_decoder(task, d_inp, vocab, embedder, args):
 class MultiTaskModel(nn.Module):
     '''
     Giant model with task-specific components and a shared word and sentence encoder.
+    This model samples the tasks passed in  
+    Pass in label.
```
Comment (W4ngatang) [inline src/models.py]: Pass in label? Not sure what this comment means

File: src/models.py [hunk 5]
```diff
@@ -141,12 +124,9 @@ def build_model(args, vocab, pretrained_embs, tasks):
...
     else:
         assert_for_log(False, "No valid sentence encoder specified.")
+    return sent_encoder 
 
-    d_sent += args.skip_embs * d_emb
-
-    # Build model and classifiers
-    model = MultiTaskModel(args, sent_encoder, vocab)
-
+def build_tasks(args, tasks):
```
Comment (W4ngatang) [inline src/models.py]: could use a more informative name, like `build_task_modules`

File: src/models.py [hunk 6]
```diff
@@ -55,48 +55,41 @@
...
-        cove_layer = None
-        embedder = OpenAIEmbedderModule(args)
-        d_emb = embedder.get_output_dim()
-    else:
-        # Default case, used for ELMo, CoVe, word embeddings, etc.
-        d_emb, embedder, cove_layer = build_embeddings(args, vocab,
-                                                       tasks, pretrained_embs)
-    d_sent = args.d_hid
-
+def build_sent_encoder(args, vocab, d_emb, tasks, embedder, cove_layer):
```
Comment (iftenney) [inline src/models.py]: FYI this will conflict with #452 (probably easier to merge that into this than vice-versa?)

File: src/models.py [hunk 7]
```diff
@@ -171,9 +174,45 @@ def build_model(args, vocab, pretrained_embs, tasks):
...
+        log.info("Using OpenAI transformer model; skipping other embedders.")
+        cove_layer = None
+        # Here, this uses openAIEmbedder.
+        embedder = OpenAIEmbedderModule(args)
+        d_emb = embedder.get_output_dim()
+    else:
+        # Default case, used for ELMo, CoVe, word embeddings, etc.
+        d_emb, embedder, cove_layer = build_embeddings(args, vocab,
+                                                       tasks, pretrained_embs)
+    d_sent = args.d_hid
```
Comment (iftenney) [inline src/models.py]: This looks like it's overwritten on the next line?

File: src/models.py [hunk 8]
```diff
@@ -171,9 +174,45 @@ def build_model(args, vocab, pretrained_embs, tasks):
...
+
+    sent_encoder, d_sent = build_sent_encoder(
+        args, vocab, d_emb, tasks, embedder, cove_layer)
+
+    d_sent += args.skip_embs * d_emb
```
Comment (iftenney) [inline src/models.py]: Comment as to what this does?

Comment (pruksmhc) [reply src/models.py]: This increases the input dimension of the task-specific module to become encoder_input_dimension + encoder_output_dimension (if args.skip_embs = 1).

File: README.md [hunk 9]
```diff
@@ -170,18 +178,19 @@ To add new tasks, you should:
...
+    * Your task implements a method ``task.process_split()`` that takes in a split of your data and produces a list of AllenNLP ``Instance``s. An ``Instance`` is a wrapper around a dictionary of ``(field_name, Field)`` pairs. ``Field``s are objects to help with data processing (indexing, padding, etc.). Each input and output should be wrapped in a field of the appropriate type (``TextField`` for text, ``LabelField`` for class labels, etc.). For MNLI, we wrap the premise and hypothesis in ``TextField``s and the label in ``LabelField``. See the [AllenNLP tutorial](https://allennlp.org/tutorials) or the examples in ``src/tasks.py``.  The names of the fields, e.g. ``input1``, can be named anything so long as the corresponding code in ``src/models.py`` (see next bullet) expects that named field. However make sure that the values to be predicted are either named ``labels`` (for classification or regression) or ``targs`` (for sequence generation)!
```
Comment (W4ngatang) [inline README.md]: nit: "produces a list" -> "produces an iterable"; this is somewhat important with really big datasets as we can't/aren't patient enough to load everything into memory at once

File: src/models.py [hunk 10]
```diff
@@ -348,9 +362,10 @@ def build_embeddings(args, vocab, tasks, pretrained_embs=None):
     assert d_emb, "You turned off all the embeddings, ya goof!"
     return d_emb, embedder, cove_layer
 
-
-def build_module(task, model, d_sent, d_emb, vocab, embedder, args):
-    ''' Build task-specific components for a task and add them to model. '''
+def build_task_specific_components(task, model, d_sent, d_emb, vocab, embedder, args):
```
Comment (W4ngatang) [inline src/models.py]: nit: could you rename this to `_build_task_module`? The name is really close to `build_task_modules` as is, and the underscore I think will make a little clearer what's going on.

File: README.md (hunk 9 above)
Comment (W4ngatang) [inline README.md]: nit: "produces a list" -> "produces an iterable"; this is somewhat important with really big datasets as we can't/aren't patient enough to load everything into memory at once

File: src/models.py (hunk 6 above)
Comment (W4ngatang) [reply src/models.py]: let's wait to merge #452 since it's approved, and then merge master into this branch

File: src/models.py [hunk 11]
```diff
@@ -57,51 +57,18 @@
...
-    # Build embeddings.
-    if args.openai_transformer:
-        # Note: incompatible with other embedders, but logic in preprocess.py
-        # should prevent these from being enabled anyway.
-        from .openai_transformer_lm.utils import OpenAIEmbedderModule
-        log.info("Using OpenAI transformer model; skipping other embedders.")
-        cove_layer = None
-        embedder = OpenAIEmbedderModule(args)
-        d_emb = embedder.get_output_dim()
-    elif args.bert_model_name:
```
[the remark this answers is not in the document]
Comment (W4ngatang) [reply src/models.py]: I think it's merged with master now.

File: src/models.py (hunk 6 above)
Comment (W4ngatang) [reply src/models.py]: Verified that master was merged into branch.

File: src/models.py [hunk 12]
```diff
@@ -157,12 +124,13 @@ def build_model(args, vocab, pretrained_embs, tasks):
...
     else:
         assert_for_log(False, "No valid sentence encoder specified.")
+    return sent_encoder, d_sent
 
-    d_sent += args.skip_embs * d_emb
-
-    # Build model and classifiers
-    model = MultiTaskModel(args, sent_encoder, vocab)
-
+def build_task_modules(args, tasks, model, d_sent, d_emb, embedder, vocab):
```
Comment (iftenney) [inline src/models.py]: nit: while we're moving things, can we either move this to below `build_model()`, or move `build_task_specific_modules()` to above this function so the control flow doesn't jump around so much in this file?

File: src/models.py [hunk 13]
```diff
@@ -554,6 +569,8 @@ def build_decoder(task, d_inp, vocab, embedder, args):
 class MultiTaskModel(nn.Module):
```
Comment (iftenney) [inline src/models.py]: Optional, but it seems like some of the other functions in this file, like `build_task_specific_modules`, might be better implemented as class methods, since they modify this object?

**Comment by iftenney** [comment]
One request, since this touches a bunch of other codepaths: can you run with `config/edgeprobe_bert.conf`  before/after to confirm?

This should work:
```
python main.py --config_file config/edgeprobe_bert.conf \
  -o "target_tasks=edges-spr2,exp_name=bert_after"
```

**Comment by sleepinyourhat** [comment]
@pruksmhc What's this waiting on?

**Comment by pruksmhc** [comment]
Was waiting on Jenkins last week - working on closing this now

**Comment by pruksmhc** [comment]
@iftenney  ran both BERT and openai and the results were the same with this change!

**Comment by pruksmhc** [comment]
@sleepinyourhat bumping this back up
